### ISSUE TITLE
pyspark - Installing JAR dependencies

### ISSUE THREAD

**Issue by andormarkus** [aws/aws-emr-containers-best-practices#10]
Hi EMR Team,

With the latest [EMR release](https://docs.aws.amazon.com/emr/latest/EMR-on-EKS-DevelopmentGuide/docker-custom-images.html), we can build custom docker images.

Installing pip packages is straight forward
```dockerfile
FROM 107292555468.dkr.ecr.eu-central-1.amazonaws.com/spark/emr-6.3.0
USER root
COPY ./ ./
RUN pip3 install --upgrade pip
RUN pip3 install --upgrade --requirement requirements.txt
USER hadoop:hadoop

```

We would like to clean up our start-job-run because it looks like this, and we would like to pre-install the jars
```json
{
  "name": "kafka",
  "virtualClusterId": "<cluster-id>",
  "executionRoleArn": "<role-arn>",
  "releaseLabel": "emr-6.3.0-latest",
  "jobDriver": {
    "sparkSubmitJobDriver": {
      "entryPoint": "s3://mop-dataeng-sandbox/kafka.py",
      "sparkSubmitParameters": "--jars https://repo1.maven.org/maven2/org/apache/spark/spark-sql-kafka-0-10_2.12/3.1.1/spark-sql-kafka-0-10_2.12-3.1.1.jar,https://repo1.maven.org/maven2/org/apache/commons/commons-pool2/2.6.2/commons-pool2-2.6.2.jar,https://repo1.maven.org/maven2/org/apache/kafka/kafka-clients/2.6.0/kafka-clients-2.6.0.jar,https://repo1.maven.org/maven2/org/apache/spark/spark-token-provider-kafka-0-10_2.12/3.1.1/spark-token-provider-kafka-0-10_2.12-3.1.1.jar,https://repo1.maven.org/maven2/org/apache/spark/spark-tags_2.12/3.1.1/spark-tags_2.12-3.1.1.jar"
    }
  },
}
```

How can we install JAR dependencies? I have tried the following solution: All of the jars file are successfully installed into the `/usr/lib/spark/jars/` directory, however spark does not pick up these new dependencies.
```dockerfile
FROM 107292555468.dkr.ecr.eu-central-1.amazonaws.com/spark/emr-6.3.0
USER root
COPY ./ ./
RUN pip3 install --upgrade pip
RUN pip3 install --upgrade --requirement requirements.txt
# Kafka
ADD https://repo1.maven.org/maven2/org/apache/spark/spark-sql-kafka-0-10_2.12/3.1.1/spark-sql-kafka-0-10_2.12-3.1.1.jar /usr/lib/spark/jars/spark-sql-kafka-0-10_2.12-3.1.1.jar
ADD https://repo1.maven.org/maven2/org/apache/commons/commons-pool2/2.6.2/commons-pool2-2.6.2.jar /usr/lib/spark/jars/commons-pool2-2.6.2.jar
ADD https://repo1.maven.org/maven2/org/apache/kafka/kafka-clients/2.6.0/kafka-clients-2.6.0.jar /usr/lib/spark/jars/kafka-clients-2.6.0.jar
ADD https://repo1.maven.org/maven2/org/apache/spark/spark-token-provider-kafka-0-10_2.12/3.1.1/spark-token-provider-kafka-0-10_2.12-3.1.1.jar /usr/lib/spark/jars/spark-token-provider-kafka-0-10_2.12-3.1.1.jar
ADD https://repo1.maven.org/maven2/org/apache/spark/spark-tags_2.12/3.1.1/spark-tags_2.12-3.1.1.jar /usr/lib/spark/jars/spark-tags_2.12-3.1.1.jar
USER hadoop:hadoop
```

**Comment by saravanan-sathyamoorthy** [comment]
@andormarkus EMR On EKS base docker image has PATH set as PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
You can either copy your dependences in any of this path or add the path(/usr/lib/spark/jars) to the PATH env variable in your docker file.

**Comment by andormarkus** [comment]
Hi @saravanan-sathyamoorthy,

I have tried all of your recommendations, but I still get this error code:


```bash
pyspark.sql.utils.AnalysisException: Failed to find data source: kafka. Please deploy the application as per the deployment section of "Structured Streaming + Kafka Integration Guide".
```

My driver logs:
[logs-from-spark-kubernetes-driver-in-spark-00000002ukruoblq7v8-driver.txt](https://github.com/aws/aws-emr-containers-best-practices/files/6803605/logs-from-spark-kubernetes-driver-in-spark-00000002ukruoblq7v8-driver.txt)

My jobrun config:
```json
{
  "name": "KafkaTester",
  "virtualClusterId": "############",
  "executionRoleArn": "arn:aws:iam::#############:role/iamrole-emrcontainers-dev",
  "releaseLabel": "emr-6.3.0-latest",
  "jobDriver": {
    "sparkSubmitJobDriver": {
      "entryPoint": "s3://my-bucket/test_kafka_job.py",
      "sparkSubmitParameters": "--conf spark.executor.instances=1 --conf spark.executor.memory=4G --conf spark.executor.cores=2 --conf spark.driver.cores=1"
    }
  },
  "configurationOverrides": {
    "applicationConfiguration": [
      {
        "classification": "spark-defaults",
        "properties": {
          "spark.kubernetes.node.selector.namespace": "emr-containers-temp",
          "spark.kubernetes.container.image": "my-image"
        }
      }
    ]
  }
}
```

**Comment by andormarkus** [comment]
Hi @saravanan-sathyamoorthy,

I have managed to solve the problem.  Based on EMR on EC2 [documentation](https://aws.amazon.com/premiumsupport/knowledge-center/emr-permanently-install-library/) any `.jar` file which is added to the `/usr/lib/spark/jars/` will be automatically installed by spark.   

In Dockerfile we are adding these file as root user and these file will get `-rw-------` permission while the original files have `-rw-r--r--` permission.  EMR on EKS uses hadoop:hadoop to run spark jobs and files with `-rw-------` permission are hidden from this user and can not be imported.   

It is a very easy fix, we need to make these file readable for all the users with the following command `chmod -R +r  /usr/lib/spark/jars` and the files will have `-rw-r--r--` permission.  

Sample Dockerfile

```Dockerfile
FROM 107292555468.dkr.ecr.eu-central-1.amazonaws.com/spark/emr-6.3.0

USER root

ARG JAR_HOME=/usr/lib/spark/jars/

# Kafka
ADD https://repo1.maven.org/maven2/org/apache/spark/spark-sql-kafka-0-10_2.12/3.1.1/spark-sql-kafka-0-10_2.12-3.1.1.jar $JAR_HOME
ADD https://repo1.maven.org/maven2/org/apache/commons/commons-pool2/2.6.2/commons-pool2-2.6.2.jar $JAR_HOME
ADD https://repo1.maven.org/maven2/org/apache/kafka/kafka-clients/2.6.0/kafka-clients-2.6.0.jar $JAR_HOME
ADD https://repo1.maven.org/maven2/org/apache/spark/spark-token-provider-kafka-0-10_2.12/3.1.1/spark-token-provider-kafka-0-10_2.12-3.1.1.jar $JAR_HOME
ADD https://repo1.maven.org/maven2/org/apache/spark/spark-tags_2.12/3.1.1/spark-tags_2.12-3.1.1.jar $JAR_HOME

RUN chmod -R +r  /usr/lib/spark/jars

USER hadoop:hadoop
```

**Comment by andormarkus** [comment]
Hi @saravanan-sathyamoorthy,

I have found another way to install jar dependencies as well. Spark runs ivy to get all of its dependencies (packages). We are executing a "dummy" run to make spark downloads its packages. These .jars are saved in `/root/.ivy2/jars/` which we can move to `/usr/lib/spark/jars/` for further use. These jars having `-rw-r--r--` permission and does not require further modifications.     

The advantage of this method is ivy download the dependencies of the package as well, and we needed to specify only `org.apache.spark:spark-sql-kafka-0-10_2.12:3.1.2` instead of 5 jars files above.

```dockerfile
FROM 107292555468.dkr.ecr.eu-central-1.amazonaws.com/spark/emr-6.3.0

USER root

ARG KAFKA_PKG="org.apache.spark:spark-sql-kafka-0-10_2.12:3.1.2"

RUN spark-submit run-example --packages $KAFKA_PKG --deploy-mode=client --master=local[1] SparkPi
RUN mv /root/.ivy2/jars/* /usr/lib/spark/jars/

USER hadoop:hadoop
```
